simulator/simulator.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// WaitForBootFinished waits until simulator finsihed boot
+func (m manager) WaitForBootFinished(id string, timeout time.Duration) error {
+	timer := time.NewTimer(timeout)
+	defer func() {
+		timer.Stop()
+	}()
+
+	launchDoneCh := make(chan error, 1)
+	doWait := func() {
+		waitCmd := m.commandFactory.Create("xcrun", []string{"simctl", "launch", id, "com.apple.Preferences"}, &command.Opts{
+			Stdout: os.Stderr,
+			Stderr: os.Stderr,
+		})
+
+		m.logger.Println()
+		m.logger.TDonef("$ %s", waitCmd.PrintableCommandArgs())
+		launchDoneCh <- waitCmd.Run()
+	}
+
+	go doWait()
+
+	for {
+		select {
+		case err := <-launchDoneCh:
+			{
+				if err != nil {
+					return fmt.Errorf("failed to wait for simulator boot: %w", err)
+				}
+				return nil // launch succeeded
+			}
+		case <-timer.C:
+			return fmt.Errorf("failed to boot Simulator in %s", timeout)
+		}
+	}
+}
+
+// EnableVerboseLog enables verbose log. Simulator needs to be booted.
+func (m manager) EnableVerboseLog(id string) error {
+	cmd := m.commandFactory.Create("xcrun", []string{"simctl", "logverbose", id, "enable"}, &command.Opts{
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+	})
+
+	m.logger.Donef("$ %s", cmd.PrintableCommandArgs())
+	if err := cmd.Run(); err != nil {
+		if errorutil.IsExitStatusError(err) {
+			m.logger.Warnf("Failed to enable Simulator verbose logging, command exited with code %d", err)
+			return nil
+		}
+
+		return fmt.Errorf("failed to enable Simulator verbose logging, command execution failed: %v", err)
+	}
+
+	return nil
+}
+
+// CollectDiagnostics collects Simulator diagnostics
+func (m manager) CollectDiagnostics() (string, error) {
+	diagnosticsName, err := m.diagnosticsName()
+	if err != nil {
+		return "", err
+	}
+
+	diagnosticsOutDir, err := ioutil.TempDir("", diagnosticsName)
+	if err != nil {
+		return "", fmt.Errorf("failed to collect Simulator diagnostics, could not create temporary directory: %v", err)
+	}
+
+	cmd := m.commandFactory.Create("xcrun", []string{"simctl", "diagnose", "-b", "--no-archive", fmt.Sprintf("--output=%s", diagnosticsOutDir)}, &command.Opts{
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+		Stdin:  bytes.NewReader([]byte("\n")),
+	})
+
+	m.logger.Donef("$ %s", cmd.PrintableCommandArgs())
+	if err := cmd.Run(); err != nil {
+		if errorutil.IsExitStatusError(err) {
+			return "", fmt.Errorf("failed to collect Simulator diagnostics: %v", err)
+
+		}
+		return "", fmt.Errorf("failed to collect Simulator diagnostics, command execution failed: %v", err)
+	}
+
+	return diagnosticsOutDir, nil
+}
+
+// Shutdown shuts down the Simulator
+func (m manager) Shutdown(id string) error {
+	cmd := m.commandFactory.Create("xcrun", []string{"simctl", "shutdown", id}, &command.Opts{
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+	})
+
+	m.logger.Donef("$ %s", cmd.PrintableCommandArgs())
+	exitCode, err := cmd.RunAndReturnExitCode()
+	if err != nil {
+		if errorutil.IsExitStatusError(err) {
+			if exitCode == 149 { // Simulator already shut down
```

Comment:
This comment is probably enough for this particular case, do you think it's reasonable to add some constants for these exit codes?

Reply:
Extracted to constant.